src/window/event_time_clock.rs

```diff
@@ -0,0 +1,277 @@
+        }
+    }
+}
+
+pub(crate) struct EventClock {
+    dt_getter: TdPyCallable,
+    late: chrono::Duration,
+    clock: InternalClock,
+    // State
+    latest_event_time: Option<DateTime<Utc>>,
+    system_time_of_last_event: Option<DateTime<Utc>>,
+    watermark: DateTime<Utc>,
+}
+
+impl EventClock {
+    fn new(
+        dt_getter: TdPyCallable,
+        late: chrono::Duration,
+        latest_event_time: Option<DateTime<Utc>>,
+        system_time_of_last_event: Option<DateTime<Utc>>,
+        watermark: DateTime<Utc>,
+        clock: InternalClock,
+    ) -> Self {
+        Self {
+            dt_getter,
+            late,
+            latest_event_time,
+            system_time_of_last_event,
+            watermark,
+            clock,
+        }
+    }
+
+    fn get_event_time(&self, event: &TdPyAny) -> DateTime<Utc> {
+        Python::with_gil(|py| {
+            self.dt_getter
+                // Call the event time getter function with the event as parameter
+                .call1(py, (event.to_object(py),))
+                .unwrap()
+                // Convert to DateTime<Utc>
+                .extract(py)
+                .unwrap()
+        })
+    }
+
+    fn advance_watermark_by_event_time(&mut self, event_time: DateTime<Utc>) {
+        // Get the latest_event_time:
+        // max between self.latest_event_time and event_time
+        // if self.latest_event_time is some, otherwise just event_time.
+        let latest_event_time = self
+            .latest_event_time
+            .map(|latest_event_time| latest_event_time.max(event_time))
+            .unwrap_or(event_time);
+        // Assign it to self.latest_event_time
+        self.latest_event_time = Some(latest_event_time);
+
+        // Get current clock's time
+        let now = self.clock.time();
+        // Get watermark
+        let watermark = now
+            .checked_add_signed(now - latest_event_time)
+            // Panic if the watermark overflowed
+            .expect("watermark in date range (overflowed)")
+            .checked_sub_signed(self.late)
+            // Panic if the watermark underflowed
+            .expect("watermark in date range (underflowed)");
+        // But only assign it to self.watermark if it's after the current
+        // one, since it can be advanced by system time too.
+        self.watermark = watermark.max(self.watermark);
+
+        // Also update self.system_time_of_last_event
+        self.system_time_of_last_event = Some(now);
+    }
+
+    fn advance_watermark_by_system_time(&mut self) {
+        let now = self.clock.time();
+        let elapsed = now - self.system_time_of_last_event.unwrap_or(self.watermark);
+        self.watermark = self
+            .watermark
+            .checked_add_signed(elapsed)
+            .expect("watermark in date range (overflowed)");
+    }
+}
+
+impl Clock<TdPyAny> for EventClock {
+    fn watermark(&mut self, next_value: &Poll<Option<TdPyAny>>) -> DateTime<Utc> {
```

Comment:
After thinking through all these other comments, perhaps a simpler way to phrase this is to store `late_time` which is a sort of "pure event time watermark", advancing when new (later) elements come in. Then always add the system time to that to generate the real watermark on the fly (since the system time is always changing on every call).

Is that generally right? Does that make things more or less clear? I think what you have here is also correct too.

```rs
match next_value {
    Poll::Ready(Some(event)) => {
        let event_late_time = self.time_for(event) - late;
        if event_late_time > self.late_time {
            self.late_time = event_late_time;
            self.system_time_of_last_event = now;
        }
    }
    Poll::Ready(None) => {
        self.late_time = MAX_UTC;
        self.system_time_of_last_event = now;
    }
    Poll::Pending => {}
}

let system_duration_since_last_event = now - self.system_time_of_last_event;
let watermark = self.late_time.checked_add_signed(system_duration_since_last_event).unwrap_or(MAX_UTC);
watermark
```

Reply:
Ok, now it is clear what we want here, fixed